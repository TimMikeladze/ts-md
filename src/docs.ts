import { z } from 'zod'
import { getDocsForDirectory, writeDocsToDirectory } from '.'

export const zDocsConfig = z.object({
  input: z.string(),
  output: z.string().default('docs'),
  extension: z.string().default('.md').optional()
})

export const docs = (config: z.infer<typeof zDocsConfig>) => {
  const data = zDocsConfig.parse(config)

  writeDocsToDirectory(
    getDocsForDirectory(data.input),
    data.output,
    data.extension
  )
}
