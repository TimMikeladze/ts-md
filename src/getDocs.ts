/* eslint-disable no-use-before-define */
import { readFileSync, readdirSync, statSync } from 'fs'
import { join } from 'path'
import ts, { SourceFile } from 'typescript'

export interface DocsForFile {
  classes: ReturnType<typeof getDocsForClasses>
  enums: ReturnType<typeof getDocsForEnums>
  interfaces: ReturnType<typeof getDocsForInterfaces>
  types: ReturnType<typeof getDocsForTypes>
}

export const getDocsForFile = (filePath: string): DocsForFile => {
  const fileContent = readFileSync(filePath, 'utf8')
  const sourceFile = ts.createSourceFile(
    filePath,
    fileContent,
    ts.ScriptTarget.ESNext,
    true
  )

  return {
    classes: getDocsForClasses(sourceFile),
    interfaces: getDocsForInterfaces(sourceFile),
    types: getDocsForTypes(sourceFile),
    enums: getDocsForEnums(sourceFile)
  }
}

export type DocsForDirectory = Map<string, DocsForFile>

export const getDocsForDirectory = (
  directoryPath: string
): DocsForDirectory => {
  const fileDocsMap = new Map<string, DocsForFile>()

  const traverseDirectory = (currentPath: string) => {
    const files = readdirSync(currentPath)

    files.forEach((file) => {
      const fullPath = join(currentPath, file)
      const stats = statSync(fullPath)

      if (stats.isDirectory()) {
        traverseDirectory(fullPath)
      } else if (file.endsWith('.ts') || file.endsWith('.tsx')) {
        const fileDocs = getDocsForFile(fullPath)
        fileDocsMap.set(fullPath, fileDocs)
      }
    })
  }

  traverseDirectory(directoryPath)

  return fileDocsMap
}

interface ClassMemberInfo {
  documentation: string
  isStatic: boolean
  kind: ts.SyntaxKind
  memberName: string
}

interface ClassInfo {
  className: string
  documentation: string
  members: ClassMemberInfo[]
  path: string
  tags: {
    name: string
  }[]
}

export const getDocsForClasses = (sourceFile: SourceFile) => {
  const classInfo: ClassInfo[] = []

  const visit = (node: ts.Node) => {
    if (ts.isClassDeclaration(node)) {
      const className = node.name ? node.name.escapedText : '<Anonymous>'
      const classDocs = ts
        .getJSDocCommentsAndTags(node)
        .map((comment) => comment.getText())
        .join('\n')

      const tags = [...ts.getJSDocTags(node)].map((tag) => ({
        name: tag.tagName.text
      }))

      classInfo.push({
        className: className.toString(),
        documentation: classDocs.trim(),
        path: sourceFile.fileName,
        tags,
        members: getDocsForClassMembers(node)
      })
    }

    ts.forEachChild(node, visit)
  }

  visit(sourceFile)

  return { info: classInfo }
}

export const getDocsForClassMembers = (
  classDeclaration: ts.ClassDeclaration
): ClassMemberInfo[] => {
  const membersInfo: ClassMemberInfo[] = []

  const visit = (node: ts.Node) => {
    if (ts.isConstructorDeclaration(node)) {
      membersInfo.push({
        memberName: 'constructor',
        documentation: ts
          .getJSDocCommentsAndTags(node)
          .map((comment) => comment.getText())
          .join('\n'),
        isStatic: false,
        kind: ts.SyntaxKind.Constructor
      })
    } else if (
      ts.isMethodDeclaration(node) ||
      ts.isGetAccessorDeclaration(node) ||
      ts.isSetAccessorDeclaration(node)
    ) {
      const memberName = node.name ? node.name.getText() : '<Unnamed>'
      const isStatic =
        node.modifiers?.some(
          (modifier) => modifier.kind === ts.SyntaxKind.StaticKeyword
        ) ?? false
      const kind = node.kind

      membersInfo.push({
        memberName,
        documentation: ts
          .getJSDocCommentsAndTags(node)
          .map((comment) => comment.getText())
          .join('\n'),
        isStatic,
        kind
      })
    } else if (ts.isPropertyDeclaration(node) || ts.isPropertySignature(node)) {
      const memberName = node.name ? node.name.getText() : '<Unnamed>'
      const isStatic =
        node.modifiers?.some(
          (modifier) => modifier.kind === ts.SyntaxKind.StaticKeyword
        ) ?? false
      const kind = node.kind

      membersInfo.push({
        memberName,
        documentation: ts
          .getJSDocCommentsAndTags(node)
          .map((comment) => comment.getText())
          .join('\n'),
        isStatic,
        kind
      })
    }

    ts.forEachChild(node, visit)
  }

  visit(classDeclaration)

  return membersInfo
}

interface InterfaceMemberInfo {
  documentation: string
  memberName: string
}

export const getDocsForInterfaceMembers = (
  interfaceDeclaration: ts.InterfaceDeclaration
): InterfaceMemberInfo[] => {
  const interfaceMemberInfo: InterfaceMemberInfo[] = []

  interfaceDeclaration.members.forEach((member) => {
    const memberName = member.name ? member.name.getText() : '<Unnamed Member>'
    const memberDocs = ts
      .getJSDocCommentsAndTags(member)
      .map((comment) => comment.getText())
      .join('\n')

    interfaceMemberInfo.push({
      memberName,
      documentation: memberDocs.trim()
    })
  })

  return interfaceMemberInfo
}

interface InterfaceInfo {
  documentation: string
  interfaceName: string
  members: InterfaceMemberInfo[]
}

export const getDocsForInterfaces = (
  sourceFile: ts.SourceFile
): InterfaceInfo[] => {
  const interfaceInfo: InterfaceInfo[] = []

  const visit = (node: ts.Node) => {
    if (ts.isInterfaceDeclaration(node)) {
      const interfaceName = node.name ? node.name.getText() : '<Unnamed>'
      const interfaceDocs = ts
        .getJSDocCommentsAndTags(node)
        .map((comment) => comment.getText())
        .join('\n')

      interfaceInfo.push({
        interfaceName,
        documentation: interfaceDocs.trim(),
        members: getDocsForInterfaceMembers(node)
      })
    }

    ts.forEachChild(node, visit)
  }

  visit(sourceFile)

  return interfaceInfo
}

interface TypeMemberInfo {
  documentation: string
  memberName: string
}

export const getDocsForTypeMembers = (
  typeAliasDeclaration: ts.TypeAliasDeclaration
): TypeMemberInfo[] => {
  const typeMemberInfo: TypeMemberInfo[] = []

  const visit = (node: ts.Node) => {
    if (ts.isPropertySignature(node) || ts.isMethodSignature(node)) {
      const memberName = node.name ? node.name.getText() : '<Unnamed Member>'
      const memberDocs = ts
        .getJSDocCommentsAndTags(node)
        .map((comment) => comment.getText())
        .join('\n')

      typeMemberInfo.push({
        memberName,
        documentation: memberDocs.trim()
      })
    }

    ts.forEachChild(node, visit)
  }

  visit(typeAliasDeclaration)

  return typeMemberInfo
}

interface TypeInfo {
  documentation: string
  members: TypeMemberInfo[]
  typeName: string
}

export const getDocsForTypes = (sourceFile: ts.SourceFile): TypeInfo[] => {
  const typeInfo: TypeInfo[] = []

  const visit = (node: ts.Node) => {
    if (ts.isTypeAliasDeclaration(node)) {
      const typeName = node.name
        ? node.name.escapedText.toString()
        : '<Unnamed Type Alias>'
      const typeDocs = ts
        .getJSDocCommentsAndTags(node)
        .map((comment) => comment.getText())
        .join('\n')

      typeInfo.push({
        typeName,
        documentation: typeDocs.trim(),
        members: getDocsForTypeMembers(node)
      })
    } else {
      ts.forEachChild(node, visit)
    }
  }

  visit(sourceFile)

  return typeInfo
}

interface EnumMemberInfo {
  documentation: string
  memberName: string
}

export const getDocsForEnumMembers = (
  enumDeclaration: ts.EnumDeclaration
): EnumMemberInfo[] => {
  const enumMemberInfo: EnumMemberInfo[] = []

  enumDeclaration.members.forEach((member) => {
    const memberName = member.name ? member.name.getText() : '<Unnamed Member>'
    const memberDocs = ts
      .getJSDocCommentsAndTags(member)
      .map((comment) => comment.getText())
      .join('\n')

    enumMemberInfo.push({
      memberName,
      documentation: memberDocs.trim()
    })
  })

  return enumMemberInfo
}

interface EnumInfo {
  documentation: string
  enumName: string
  members: EnumMemberInfo[]
}

export const getDocsForEnums = (sourceFile: ts.SourceFile): EnumInfo[] => {
  const enumInfo: EnumInfo[] = []

  const visit = (node: ts.Node) => {
    if (ts.isEnumDeclaration(node)) {
      const enumName = node.name
        ? node.name.escapedText.toString()
        : '<Unnamed Enum>'
      const enumDocs = ts
        .getJSDocCommentsAndTags(node)
        .map((comment) => comment.getText())
        .join('\n')

      enumInfo.push({
        enumName,
        documentation: enumDocs.trim(),
        members: getDocsForEnumMembers(node)
      })
    } else {
      ts.forEachChild(node, visit)
    }
  }

  visit(sourceFile)

  return enumInfo
}
