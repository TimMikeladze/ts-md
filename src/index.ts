export * from './getDocs'
export * from './writeDocs'
export * from './docs'
