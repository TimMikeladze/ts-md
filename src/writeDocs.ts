import { basename, dirname, extname, join } from 'path'
import { DocsForDirectory, DocsForFile } from './getDocs'
import { mkdirSync, writeFileSync } from 'fs'

const convertToMarkdown = (docs: DocsForFile): string => {
  return `# Classes
${docs.classes.info
  .map((cls) => `## ${cls.className}\n${cls.documentation}`)
  .join('\n\n')}

# Enums
${docs.enums
  .map((enumInfo) => `## ${enumInfo.enumName}\n${enumInfo.documentation}`)
  .join('\n\n')}

# Interfaces
${docs.interfaces
  .map(
    (interfaceInfo) =>
      `## ${interfaceInfo.interfaceName}\n${interfaceInfo.documentation}`
  )
  .join('\n\n')}

# Types
${docs.types
  .map((typeInfo) => `## ${typeInfo.typeName}\n${typeInfo.documentation}`)
  .join('\n\n')}`
}

export const writeDocsToFile = (
  docs: DocsForFile,
  filePath: string,
  extension: string = '.md'
) => {
  const markdownContent = convertToMarkdown(docs)
  const outputFileName = basename(filePath, extname(filePath)) + extension
  const outputPath = join(dirname(filePath), outputFileName)

  mkdirSync(dirname(outputPath), { recursive: true })

  writeFileSync(outputPath, markdownContent, 'utf8')
}

export const writeDocsToDirectory = (
  docsForDirectory: DocsForDirectory,
  directoryPath: string,
  extension: string = '.md'
) => {
  mkdirSync(directoryPath, { recursive: true })

  docsForDirectory.forEach((docsForFile, filePath) => {
    writeDocsToFile(
      docsForFile,
      filePath.replace('tests/assets', directoryPath),
      extension
    )
  })
}
