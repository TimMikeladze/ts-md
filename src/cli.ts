import { Command } from 'commander'
import { getDocsForDirectory } from './getDocs'
import { writeDocsToDirectory } from './writeDocs'
;(async () => {
  const program = new Command()

  program.name('ts-md').description('').version('x.x.x')

  program.option('-i --input <inputPath>', 'Path to input directory')

  program.option('-o --output <outputPath>', 'Path to output directory')

  program.option(
    '-e --extension <extension>',
    'File extension to output, defaults to .md'
  )

  program
    .command('docs')
    .description('Generate documentation')
    .action(async () => {
      if (!program.opts().input) {
        console.error('Error: input path is required')
        return
      }
      if (!program.opts().output) {
        console.error('Error: output path is required')
      }
      const extension = program.opts().extension || '.md'
      const inputPath = program.opts().input
      const outputPath = program.opts().output

      writeDocsToDirectory(
        getDocsForDirectory(inputPath),
        outputPath,
        extension
      )
    })

  const defaultCommand = async () => {}

  try {
    await program.parseAsync(process.argv)
    if (process.argv.length === 2) {
      await defaultCommand()
    }
  } catch (err: any) {
    console.error(err.message)
  }
})()
