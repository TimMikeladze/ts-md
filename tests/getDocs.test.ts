import { expect, it } from 'vitest'
import { getDocsForDirectory } from '../src/getDocs'

it('getDocsForDirectory', () => {
  const res = getDocsForDirectory('tests/assets')

  expect(res).toMatchSnapshot()
})
