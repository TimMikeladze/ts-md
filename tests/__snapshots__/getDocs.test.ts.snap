// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getDocsForDirectory 1`] = `
Map {
  "tests/assets/Example.ts" => {
    "classes": {
      "info": [
        {
          "className": "BasicClass",
          "documentation": "/**
 * This is a basic class.
 *
 * @export
 * @class BasicClass
 * @example
 * const basicClass = new BasicClass()
 * basicClass.hello()
 */",
          "members": [
            {
              "documentation": "/**
   * This is a constructor.
   * @memberof BasicClass
   * @example
   * const basicClass = new BasicClass()
   */",
              "isStatic": false,
              "kind": 175,
              "memberName": "constructor",
            },
            {
              "documentation": "/**
   * This is a basic method.
   *
   * @memberof BasicClass
   */",
              "isStatic": false,
              "kind": 173,
              "memberName": "hello",
            },
          ],
          "path": "tests/assets/Example.ts",
          "tags": [
            {
              "name": "export",
            },
            {
              "name": "class",
            },
            {
              "name": "example",
            },
          ],
        },
        {
          "className": "AnotherClass",
          "documentation": "",
          "members": [
            {
              "documentation": "/**
   * This is a constructor.
   * @memberof AnotherClass
   * @example
   * const anotherClass = new AnotherClass()
   */",
              "isStatic": true,
              "kind": 171,
              "memberName": "foo",
            },
            {
              "documentation": "",
              "isStatic": false,
              "kind": 171,
              "memberName": "bar",
            },
            {
              "documentation": "",
              "isStatic": false,
              "kind": 175,
              "memberName": "constructor",
            },
            {
              "documentation": "",
              "isStatic": true,
              "kind": 173,
              "memberName": "hello",
            },
          ],
          "path": "tests/assets/Example.ts",
          "tags": [],
        },
        {
          "className": "AbstractClass",
          "documentation": "/**
 * This is an abstract class.
 * @abstract
 */",
          "members": [
            {
              "documentation": "/**
   * This is a constructor.
   * @memberof AbstractClass
   * @example
   * const abstractClass = new AbstractClass()
   */",
              "isStatic": false,
              "kind": 175,
              "memberName": "constructor",
            },
            {
              "documentation": "",
              "isStatic": false,
              "kind": 173,
              "memberName": "hello",
            },
          ],
          "path": "tests/assets/Example.ts",
          "tags": [
            {
              "name": "abstract",
            },
          ],
        },
      ],
    },
    "enums": [
      {
        "documentation": "/**
 * This is an enum.
 */",
        "enumName": "Enum",
        "members": [
          {
            "documentation": "/**
   * This is a property.
   */",
            "memberName": "bar",
          },
          {
            "documentation": "",
            "memberName": "foo",
          },
        ],
      },
    ],
    "interfaces": [
      {
        "documentation": "/**
 * This is an interface.
 */",
        "interfaceName": "Interface",
        "members": [
          {
            "documentation": "",
            "memberName": "bar",
          },
          {
            "documentation": "/**
   * This is a property.
   */",
            "memberName": "foo",
          },
        ],
      },
    ],
    "types": [
      {
        "documentation": "/**
 * This is a type
 */",
        "members": [
          {
            "documentation": "",
            "memberName": "bar",
          },
          {
            "documentation": "/**
   * This is a property.
   */",
            "memberName": "foo",
          },
        ],
        "typeName": "Type",
      },
    ],
  },
}
`;
